### DISCUSSION TITLE
yarn build failed for Javascript setup with error "Unrecognized Token"

### DISCUSSION THREAD

**Discussion by hemantbaxi** [perspective-dev/perspective#1428] [Q&A]
## Bug Report
With perspective master branch,  it builds fine in python setup but failing to build in Javascript setup with following error:

### Steps to Reproduce:
 1. Clone the master branch from github
 2. select Javascript setup
 3. yarn build ... I did get some other errors before getting to this which were because of missing dependencies and got fixed after I installed them.

### Expected Result:
Build to work fine.

### Actual Result:
```
$ babel src/js --source-maps --out-dir dist/esm
SyntaxError: src/js/api/client.js: Unexpected token (61:83)
  59 |                 this.send(msg);
  60 | 
> 61 |                 if ((msg.cmd === "table" || msg.cmd === "view") && !this._features?.wait_for_response && resolve) {
     |                                                                                    ^
  62 |                     resolve();
  63 |                 }
  64 |             });
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "build:babel:esm" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn build exited 1 in '@finos/perspective'
lerna ERR! yarn build exited 1 in '@finos/perspective'
```

After doing google search, it seems this is babel optional-chaining feature which webpack is not recognizing. I checked 
I have @babel/preset-env plugin. These are the versions of babel and webpack on my system

axihemant@hb-finos:~/finos/perspective$ npm list webpack
/home/baxihemant/finos/perspective
└── webpack@4.46.0 

baxihemant@hb-finos:~/finos/perspective$ npm list babel
/home/baxihemant/finos/perspective
└── babel@6.23.0 

baxihemant@hb-finos:~/finos/perspective$ npm list babel-cli
/home/baxihemant/finos/perspective
└── babel-cli@6.26.0 

baxihemant@hb-finos:~/finos/perspective$ npm list @babel/preset-env
/home/baxihemant/finos/perspective
└── @babel/preset-env@7.14.2 

### Environment:
master branch
OS : Ubuntu 20.10
npm -v
6.14.8
Virtualized environment

Thanks,
Hemant

**Comment by texodus** [comment]
There are no missing dependencies, [the build works fine](https://dev.azure.com/finosfoundation/perspective/_build/results?buildId=2623).  You likely did not run `yarn` first.

**Comment by hemantbaxi** [comment]
Thanks.
Just doing yarn wasn't enough. I looked at the commands which CI was running for setup and dependencies. Following are the steps that finally gave me successful js build after doing a fresh clone of the repo.

```
yarn
python -m pip install --upgrade pip wheel setuptools jupyterlab numpy "pyarrow>=2" "black==20.8b1" flake8-black
  --> failed for testresources, so had to first install that and ran this again
python -m pip install --upgrade testresources
python -m pip install --upgrade pip wheel setuptools jupyterlab numpy "pyarrow>=2" "black==20.8b1" flake8-black
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
sudo apt-get install rust
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
curl --proto '=https' --tlsv1.2 https://sh.rustup.rs -sSf | sh
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
source $HOME/.cargo/env
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
sudo node scripts/install_tools.js
yarn build_js
```

However, now it is failing for
**_yarn test_js_**

```
lerna info Executing command in 1 package: "yarn test"
$ yarn test:build && wasm-pack test --chrome --headless && yarn test:clean
$ cpx ../../packages/perspective/dist/umd/perspective.inline.js pkg/
[INFO]: Checking for the Wasm target...
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
[INFO]: Installing wasm-bindgen...
    Finished test [unoptimized + debuginfo] target(s) in 0.05s
     Running unittests (target/wasm32-unknown-unknown/debug/deps/perspective_vieux-e9dfafb26778469d.wasm)
Set timeout to 20 seconds...
Running headless tests in Chrome on `http://127.0.0.1:37057/`
Try find `webdriver.json` for configure browser's capabilities:
Not found
driver status: signal: 9                          
driver stdout:
    Starting ChromeDriver 90.0.4430.24 (4c6d850f087da467d926e8eddb76550aed655991-refs/branch-heads/4430@{#429}) on port 37057
    Only local connections are allowed.
    Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
    ChromeDriver was started successfully.

Error: non-200 response code: 404                 
{"value":{"error":"invalid session id","message":"invalid session id","stacktrace":"#0 0x55b55ebbbe89 \u003Cunknown>\n"}}
error: test failed, to rerun pass '--lib'
Error: Running Wasm tests with wasm-bindgen-test failed
Caused by: failed to execute `cargo test`: exited with exit code: 1
  full command: "cargo" "test" "--target" "wasm32-unknown-unknown"
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn test exited 1 in '@finos/perspective-vieux'
lerna ERR! yarn test exited 1 in '@finos/perspective-vieux'
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Command failed: yarn lerna --scope=@finos/perspective-vieux exec yarn test
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```
I do have webdriver installed, but it is picking up ChromeDriver...  I see that in CI, it is calling webdriver.. Not sure what is missing here.

Thanks,
Hemant

**Reply by texodus** [reply]
`chromedriver` is supposed to be downloaded when needed by `puppeteer`/`wasm-pack`.

I believe I may have seen this particular failure or something similar when the local `chromedriver` does not match the local `chromium`, which I resovled by [downloading an updated `chromedriver` for my platform manually ala here](https://github.com/rustwasm/wasm-pack/issues/611#issuecomment-522093207).
